### PR TITLE
[SDK] GetApplications SDK Method

### PR DESCRIPTION
**Pull request by ezeike**
## Summary

Add a GetApplications method to the Poktroll SDK. This allows clients to request a list of all staked applications.
This method takes a gateway address as a parameter and returns applications delegated to that gateway.

### Summary generated by Reviewpad on 01 Feb 24 01:23 UTC
This pull request adds the following changes:
- In `pkg/client/interface.go`, a new method `GetAllApplications` is added to the `ApplicationQueryClient` interface, which queries all on-chain applications.
- In `pkg/client/query/appquerier.go`, the implementation of the `GetAllApplications` method is added to the `appQuerier` struct, which queries all staked applications.
- In `pkg/sdk/application.go`, a new file `application.go` is added, which contains the implementation of the `GetApplications` method. This method queries a list of all on-chain applications, optionally filtered by delegated gateway.
- In `pkg/sdk/errors.go`, a new error `ErrSDKGetApplications` is added.
- In `pkg/sdk/interface.go`, the `POKTRollSDK` interface is updated to include the `GetApplications` method.
- In `pkg/sdk/sdk.go`, the `poktrollSDK` struct is updated to include the `applicationQuerier` field and initialized in the `NewPOKTRollSDK` function.

These changes enhance the SDK by adding the ability to query all on-chain applications and retrieve applications filtered by delegated gateway.

## Issue

Grove Gateway requires a list of all staked applications in order to perform QoS. 

- #324 

## Type of change

Select one or more:

- [X] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [X] **Run all unit tests**: `make go_develop_and_test`
- [X] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [X] I have tested my changes using the available tooling
- [X] I have performed a self-review of my own code
- [X] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/client/interface.go
Comment (red-0ne) [inline pkg/client/interface.go:242]: Could you add a godoc comment here?

File: pkg/sdk/application.go
Comment (red-0ne) [inline pkg/sdk/application.go:11]: I can't see how this is relevant nor how it's used.

Should we just remove this struct?

File: pkg/sdk/application.go
Comment (red-0ne) [inline pkg/sdk/application.go:16]: I don't think this comment reflects the purpose of the function. Please update it.

File: pkg/sdk/application.go
Comment (red-0ne) [inline pkg/sdk/application.go:28]: Add a `TODO_TECHDEBT` explaining that this does not scale well when the staking applications grows and that we should consider a better scaling solution.

File: pkg/sdk/interface.go
Comment (red-0ne) [inline pkg/sdk/interface.go:14]:
```suggestion
	// GetApplications returns all the existing applications that staked on the Pocket network
```

File: pkg/sdk/application.go
Comment (red-0ne) [inline pkg/sdk/application.go:19]: WDYT of renaming this to `GetDelegatingApplycationsByGateway` or `GetApplicationsDelegatingToGateway`?

cc @Olshansk

File: pkg/sdk/application.go
Comment (red-0ne) [inline pkg/sdk/application.go:21]: Could you replace the `filtered` variable name by a more contextual one? I would suggest `delegatingApplications`

cc @Olshansk , a better suggestion?

**Review by red-0ne — CHANGES_REQUESTED** [review]
Glad to see you contributing to the SDK.

I left some change requests PTAL.

**Comment by Olshansk** [comment]
@red-0ne @ezeike I'm holding off on reviewing it until we decide on [1] as outlined in [2]. Lmk what you think/find

[1]  https://github.com/pokt-network/poktroll/pull/344
[2] https://github.com/pokt-network/poktroll/issues/324

File: pkg/sdk/application.go
Comment (ezeike) [reply pkg/sdk/application.go]: That sounds alright with me.

We should decide whether querying all on-chain applications should be a part of an SDK meant to develop gateways that would have no interest in other applications.

For the moment I added a options struct for this method.

File: pkg/sdk/application.go
Comment (ezeike) [reply pkg/sdk/application.go]: Done.

File: pkg/sdk/application.go
Comment (Olshansk) [inline pkg/sdk/application.go:16]:
```suggestion
// GetApplications queries a list of all on-chain applications, filtered by the
// options provided
```

File: pkg/sdk/application.go
Comment (Olshansk) [inline pkg/sdk/application.go:18]:
```suggestion
```

File: pkg/sdk/application.go
Comment (Olshansk) [inline pkg/sdk/application.go:30]:
```suggestion
	// TODO_CONSIDERATION: Look into updating the on-chain `QueryAllApplicationRequest` for filtering
	// options to avoid needing to do it on the client side.
	for _, app := range applications {
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Review by Olshansk — APPROVED** [review]
@ezeike LGTM, but wanted to mention two things.

1. Please use the `sqasuh & merge` function before merging (after E2E tests have passed).
2. Make sure to manually update the description when doing (1)
3. Please wait for an approval from @red-0ne as well before merging it in.

**Review by red-0ne — APPROVED** [review]
LGTM!

Thanks for putting this together @ezeike.

Don't forget to add the `devnet-test-e2e` label to the PR. And do `make trigger_ci` from your local repo to make sure E2E tests are ran in the CI. :pray:

**Pull request merged**
